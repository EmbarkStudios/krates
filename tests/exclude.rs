mod util;

use krates::PkgSpec;
use util::build;

#[test]
fn excludes_workspace_member() {
    let mut kb = krates::Builder::new();
    kb.exclude(std::iter::once("a".parse::<PkgSpec>().unwrap()));

    let grafs = build("all-features.json", kb).unwrap();

    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn excludes_dependencies() {
    let mut kb = krates::Builder::new();

    let pkg_ids = [
        "bitflags",
        "bumpalo:3.11.0",
        "https://github.com/rust-lang/crates.io-index#byteorder",
        "https://github.com/rust-lang/crates.io-index#ring:0.16.20",
        "https://github.com/alexcrichton/cc-rs#cc",
    ];

    kb.exclude(pkg_ids.iter().map(|id| id.parse::<PkgSpec>().unwrap()));

    let grafs = build("all-features.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn no_roots() {
    let mut kb = krates::Builder::new();

    // To ease testing, we just remove leaves that have no dependencies
    // themselves
    let pkg_ids = ["a", "b", "c"];

    kb.exclude(pkg_ids.iter().map(|id| id.parse::<PkgSpec>().unwrap()));

    let contents = std::fs::read_to_string("tests/all-features.json")
        .map_err(|e| format!("failed to load metadata file: {}", e))
        .unwrap();

    let md: krates::cm::Metadata = serde_json::from_str(&contents)
        .map_err(|e| format!("failed to deserialize metadata: {}", e))
        .unwrap();

    match kb.build_with_metadata::<util::JustId, krates::Edge, _>(md, |_f: krates::cm::Package| {
        panic!("shouldn't get here")
    }) {
        Err(krates::Error::NoRootKrates) => {}
        _ => panic!("expected no root crates error!"),
    }
}
