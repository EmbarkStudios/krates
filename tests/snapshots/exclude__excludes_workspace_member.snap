---
source: tests/exclude.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate aho-corasick 0.7.6" ]
    1 [ label = "crate anyhow 1.0.26" ]
    2 [ label = "crate b 0.1.0 path+file:///krates/tests/ws/b" ]
    3 [ label = "crate bindgen 0.51.1" ]
    4 [ label = "crate bitflags 1.2.1" ]
    5 [ label = "crate bumpalo 3.1.2" ]
    6 [ label = "crate byteorder 1.3.2" ]
    7 [ label = "crate c 0.1.0 path+file:///krates/tests/ws/c" ]
    8 [ label = "crate cc 1.0.50" ]
    9 [ label = "crate cc 1.0.84 git+https://github.com/alexcrichton/cc-rs?branch=main" ]
    10 [ label = "crate cc 1.0.84 git+https://github.com/alexcrichton/cc-rs?rev=34d4ce4" ]
    11 [ label = "crate cexpr 0.3.6" ]
    12 [ label = "crate cfg-if 0.1.10" ]
    13 [ label = "crate clang-sys 0.28.1" ]
    14 [ label = "crate coreaudio-rs 0.9.1" ]
    15 [ label = "crate coreaudio-sys 0.2.3" ]
    16 [ label = "crate difference 2.0.0" ]
    17 [ label = "crate glob 0.3.0" ]
    18 [ label = "crate heck 0.3.1" ]
    19 [ label = "crate js-sys 0.3.35" ]
    20 [ label = "crate lazy_static 1.4.0" ]
    21 [ label = "crate leftpad 0.2.0" ]
    22 [ label = "crate libc 0.2.66" ]
    23 [ label = "crate libloading 0.5.2" ]
    24 [ label = "crate log 0.4.8" ]
    25 [ label = "crate memchr 2.2.1" ]
    26 [ label = "crate nix 0.16.1" ]
    27 [ label = "crate nom 4.2.3" ]
    28 [ label = "crate peeking_take_while 0.1.2" ]
    29 [ label = "crate proc-macro2 1.0.7" ]
    30 [ label = "crate quote 1.0.2" ]
    31 [ label = "crate regex 1.3.3" ]
    32 [ label = "crate regex-syntax 0.6.13" ]
    33 [ label = "crate ring 0.16.9" ]
    34 [ label = "crate rustc-hash 1.0.1" ]
    35 [ label = "crate shlex 0.1.1" ]
    36 [ label = "crate sourcefile 0.1.4" ]
    37 [ label = "crate spin 0.5.2" ]
    38 [ label = "crate syn 1.0.13" ]
    39 [ label = "crate thread_local 1.0.0" ]
    40 [ label = "crate unicode-segmentation 1.6.0" ]
    41 [ label = "crate unicode-xid 0.2.0" ]
    42 [ label = "crate untrusted 0.7.0" ]
    43 [ label = "crate version_check 0.1.5" ]
    44 [ label = "crate void 1.0.2" ]
    45 [ label = "crate wasm-bindgen 0.2.58" ]
    46 [ label = "crate wasm-bindgen-backend 0.2.58" ]
    47 [ label = "crate wasm-bindgen-futures 0.4.8" ]
    48 [ label = "crate wasm-bindgen-macro 0.2.58" ]
    49 [ label = "crate wasm-bindgen-macro-support 0.2.58" ]
    50 [ label = "crate wasm-bindgen-shared 0.2.58" ]
    51 [ label = "crate wasm-bindgen-webidl 0.2.58" ]
    52 [ label = "crate web-sys 0.3.35" ]
    53 [ label = "crate weedle 0.10.0" ]
    54 [ label = "crate winapi 0.2.8" ]
    55 [ label = "crate winapi 0.3.8" ]
    56 [ label = "crate winapi-i686-pc-windows-gnu 0.4.0" ]
    57 [ label = "crate winapi-x86_64-pc-windows-gnu 0.4.0" ]
    58 [ label = "feature use_std" ]
    59 [ label = "feature default" ]
    60 [ label = "feature default" ]
    61 [ label = "feature default" ]
    62 [ label = "feature runtime" ]
    63 [ label = "feature clang_6_0" ]
    64 [ label = "feature default" ]
    65 [ label = "feature default" ]
    66 [ label = "feature default" ]
    67 [ label = "feature Crypto" ]
    68 [ label = "feature Window" ]
    69 [ label = "feature verbose-errors" ]
    70 [ label = "feature default" ]
    71 [ label = "feature audio_toolbox" ]
    72 [ label = "feature audio_unit" ]
    73 [ label = "feature core_audio" ]
    74 [ label = "feature core_midi" ]
    75 [ label = "feature open_al" ]
    76 [ label = "feature default" ]
    77 [ label = "feature winerror" ]
    78 [ label = "feature errhandlingapi" ]
    79 [ label = "feature libloaderapi" ]
    80 [ label = "feature extra_traits" ]
    81 [ label = "feature default" ]
    82 [ label = "feature default" ]
    83 [ label = "feature default" ]
    84 [ label = "feature proc-macro" ]
    85 [ label = "feature default" ]
    86 [ label = "feature default" ]
    87 [ label = "feature unicode-age" ]
    88 [ label = "feature unicode-bool" ]
    89 [ label = "feature unicode-case" ]
    90 [ label = "feature unicode-gencat" ]
    91 [ label = "feature unicode-perl" ]
    92 [ label = "feature unicode-script" ]
    93 [ label = "feature unicode-segment" ]
    94 [ label = "feature ntsecapi" ]
    95 [ label = "feature wtypesbase" ]
    96 [ label = "feature default" ]
    97 [ label = "feature proc-macro" ]
    98 [ label = "feature spans" ]
    99 [ label = "feature default" ]
    100 [ label = "feature default" ]
    101 [ label = "feature default" ]
    102 [ label = "feature full" ]
    103 [ label = "feature default" ]
    104 [ label = "feature MessageEvent" ]
    105 [ label = "feature Worker" ]
    106 [ label = "feature spans" ]
    107 [ label = "feature visit" ]
    108 [ label = "feature spans" ]
    109 [ label = "feature default" ]
    110 [ label = "feature std" ]
    111 [ label = "feature std" ]
    112 [ label = "feature std" ]
    113 [ label = "feature leftpad" ]
    114 [ label = "feature leftier-strings" ]
    115 [ label = "feature lazy_static" ]
    116 [ label = "feature libloading" ]
    117 [ label = "feature gte_clang_6_0" ]
    118 [ label = "feature gte_clang_5_0" ]
    119 [ label = "feature gte_clang_4_0" ]
    120 [ label = "feature gte_clang_3_9" ]
    121 [ label = "feature gte_clang_3_8" ]
    122 [ label = "feature gte_clang_3_7" ]
    123 [ label = "feature gte_clang_3_6" ]
    124 [ label = "feature open_al" ]
    125 [ label = "feature audio_toolbox" ]
    126 [ label = "feature audio_unit" ]
    127 [ label = "feature core_audio" ]
    128 [ label = "feature core_midi" ]
    129 [ label = "feature std" ]
    130 [ label = "feature alloc" ]
    131 [ label = "feature std" ]
    132 [ label = "feature unicode-segment" ]
    133 [ label = "feature unicode-script" ]
    134 [ label = "feature unicode-perl" ]
    135 [ label = "feature unicode-gencat" ]
    136 [ label = "feature unicode-case" ]
    137 [ label = "feature unicode-bool" ]
    138 [ label = "feature unicode-age" ]
    139 [ label = "feature unicode" ]
    140 [ label = "feature thread_local" ]
    141 [ label = "feature std" ]
    142 [ label = "feature perf-literal" ]
    143 [ label = "feature aho-corasick" ]
    144 [ label = "feature memchr" ]
    145 [ label = "feature perf-inline" ]
    146 [ label = "feature perf-dfa" ]
    147 [ label = "feature perf-cache" ]
    148 [ label = "feature perf" ]
    149 [ label = "feature lazy_static" ]
    150 [ label = "feature dev_urandom_fallback" ]
    151 [ label = "feature alloc" ]
    152 [ label = "feature quote" ]
    153 [ label = "feature proc-macro" ]
    154 [ label = "feature printing" ]
    155 [ label = "feature parsing" ]
    156 [ label = "feature derive" ]
    157 [ label = "feature clone-impls" ]
    158 [ label = "feature std" ]
    159 [ label = "feature std" ]
    160 [ label = "feature spans" ]
    0 -> 25 [ label = "" ]
    0 -> 58 [ label = "" ]
    2 -> 59 [ label = "" ]
    2 -> 10 [ label = "(build)" ]
    2 -> 60 [ label = "(dev) 'cfg(target_arch = \"x86_64\")'" ]
    2 -> 47 [ label = " 'cfg(all(target_arch = \"wasm32\", target_feature = \"atomics\"))'" ]
    2 -> 47 [ label = " 'cfg(all(target_vendor = \"xboxone\"))'" ]
    3 -> 61 [ label = "" ]
    3 -> 11 [ label = "" ]
    3 -> 12 [ label = "" ]
    3 -> 62 [ label = "" ]
    3 -> 63 [ label = "" ]
    3 -> 20 [ label = "" ]
    3 -> 28 [ label = "" ]
    3 -> 29 [ label = "" ]
    3 -> 30 [ label = "" ]
    3 -> 64 [ label = "" ]
    3 -> 34 [ label = "" ]
    3 -> 35 [ label = "" ]
    7 -> 9 [ label = "(build)" ]
    7 -> 65 [ label = " 'x86_64-apple-darwin'" ]
    7 -> 66 [ label = "(dev)" ]
    7 -> 20 [ label = " 'cfg(any(target_os = \"android\", target_os = \"freebsd\", target_os = \"linux\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    7 -> 21 [ label = "" ]
    7 -> 22 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    7 -> 26 [ label = " 'x86_64-unknown-linux-gnu'" ]
    7 -> 37 [ label = " 'cfg(all(any(target_arch = \"aarch64\", target_arch = \"arm\", target_arch = \"x86\", target_arch = \"x86_64\"), not(target_os = \"ios\")))'" ]
    7 -> 67 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    7 -> 68 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    7 -> 54 [ label = " 'cfg(target_os = \"windows\")'" ]
    9 -> 22 [ label = " 'cfg(unix)'" ]
    10 -> 22 [ label = " 'cfg(unix)'" ]
    11 -> 69 [ label = "" ]
    11 -> 70 [ label = "" ]
    13 -> 17 [ label = "" ]
    13 -> 17 [ label = "(build)" ]
    13 -> 22 [ label = "" ]
    13 -> 23 [ label = "" ]
    14 -> 61 [ label = "" ]
    14 -> 15 [ label = "" ]
    14 -> 71 [ label = "" ]
    14 -> 72 [ label = "" ]
    14 -> 73 [ label = "" ]
    14 -> 74 [ label = "" ]
    14 -> 75 [ label = "" ]
    15 -> 3 [ label = "(build)" ]
    18 -> 40 [ label = "" ]
    19 -> 76 [ label = "" ]
    23 -> 8 [ label = "(build)" ]
    23 -> 77 [ label = " 'cfg(windows)'" ]
    23 -> 78 [ label = " 'cfg(windows)'" ]
    23 -> 79 [ label = " 'cfg(windows)'" ]
    24 -> 12 [ label = "" ]
    26 -> 61 [ label = "" ]
    26 -> 8 [ label = "(build) 'cfg(target_os = \"dragonfly\")'" ]
    26 -> 12 [ label = "" ]
    26 -> 80 [ label = "" ]
    26 -> 81 [ label = "" ]
    26 -> 82 [ label = "" ]
    27 -> 25 [ label = "" ]
    27 -> 58 [ label = "" ]
    27 -> 43 [ label = "(build)" ]
    29 -> 83 [ label = "" ]
    30 -> 29 [ label = "" ]
    30 -> 84 [ label = "" ]
    31 -> 85 [ label = "" ]
    31 -> 86 [ label = "" ]
    31 -> 32 [ label = "" ]
    31 -> 87 [ label = "" ]
    31 -> 88 [ label = "" ]
    31 -> 89 [ label = "" ]
    31 -> 90 [ label = "" ]
    31 -> 91 [ label = "" ]
    31 -> 92 [ label = "" ]
    31 -> 93 [ label = "" ]
    31 -> 39 [ label = "" ]
    33 -> 8 [ label = "(build)" ]
    33 -> 20 [ label = " 'cfg(any(target_os = \"android\", target_os = \"freebsd\", target_os = \"linux\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    33 -> 22 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    33 -> 37 [ label = " 'cfg(all(any(target_arch = \"aarch64\", target_arch = \"arm\", target_arch = \"x86\", target_arch = \"x86_64\"), not(target_os = \"ios\")))'" ]
    33 -> 42 [ label = "" ]
    33 -> 67 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    33 -> 68 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    33 -> 94 [ label = " 'cfg(target_os = \"windows\")'" ]
    33 -> 95 [ label = " 'cfg(target_os = \"windows\")'" ]
    34 -> 96 [ label = "" ]
    38 -> 29 [ label = "" ]
    38 -> 84 [ label = "" ]
    38 -> 30 [ label = "" ]
    38 -> 97 [ label = "" ]
    38 -> 83 [ label = "" ]
    39 -> 20 [ label = "" ]
    45 -> 12 [ label = "" ]
    45 -> 48 [ label = "" ]
    45 -> 98 [ label = "" ]
    46 -> 99 [ label = "" ]
    46 -> 20 [ label = "" ]
    46 -> 24 [ label = "" ]
    46 -> 100 [ label = "" ]
    46 -> 101 [ label = "" ]
    46 -> 102 [ label = "" ]
    46 -> 103 [ label = "" ]
    46 -> 50 [ label = "" ]
    47 -> 12 [ label = "" ]
    47 -> 19 [ label = "" ]
    47 -> 76 [ label = "" ]
    47 -> 104 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    47 -> 105 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    48 -> 101 [ label = "" ]
    48 -> 49 [ label = "" ]
    48 -> 106 [ label = "" ]
    49 -> 100 [ label = "" ]
    49 -> 101 [ label = "" ]
    49 -> 107 [ label = "" ]
    49 -> 103 [ label = "" ]
    49 -> 46 [ label = "" ]
    49 -> 108 [ label = "" ]
    49 -> 50 [ label = "" ]
    51 -> 109 [ label = "" ]
    51 -> 18 [ label = "" ]
    51 -> 24 [ label = "" ]
    51 -> 100 [ label = "" ]
    51 -> 101 [ label = "" ]
    51 -> 102 [ label = "" ]
    51 -> 103 [ label = "" ]
    51 -> 46 [ label = "" ]
    51 -> 53 [ label = "" ]
    52 -> 109 [ label = "(build)" ]
    52 -> 19 [ label = "" ]
    52 -> 36 [ label = "(build)" ]
    52 -> 76 [ label = "" ]
    52 -> 51 [ label = "(build)" ]
    53 -> 70 [ label = "" ]
    55 -> 56 [ label = " 'i686-pc-windows-gnu'" ]
    55 -> 57 [ label = " 'x86_64-pc-windows-gnu'" ]
    110 -> 0 [ label = "" ]
    110 -> 58 [ label = "" ]
    85 -> 0 [ label = "" ]
    85 -> 110 [ label = "" ]
    111 -> 1 [ label = "" ]
    109 -> 1 [ label = "" ]
    109 -> 111 [ label = "" ]
    61 -> 4 [ label = "" ]
    99 -> 5 [ label = "" ]
    112 -> 6 [ label = "" ]
    96 -> 6 [ label = "" ]
    96 -> 112 [ label = "" ]
    113 -> 7 [ label = "" ]
    113 -> 21 [ label = "" ]
    114 -> 7 [ label = "" ]
    114 -> 113 [ label = "" ]
    115 -> 7 [ label = "" ]
    115 -> 20 [ label = "" ]
    59 -> 7 [ label = "" ]
    59 -> 113 [ label = "" ]
    62 -> 13 [ label = "" ]
    62 -> 116 [ label = "" ]
    116 -> 13 [ label = "" ]
    116 -> 23 [ label = "" ]
    117 -> 13 [ label = "" ]
    118 -> 13 [ label = "" ]
    119 -> 13 [ label = "" ]
    120 -> 13 [ label = "" ]
    121 -> 13 [ label = "" ]
    122 -> 13 [ label = "" ]
    123 -> 13 [ label = "" ]
    63 -> 13 [ label = "" ]
    63 -> 123 [ label = "" ]
    63 -> 122 [ label = "" ]
    63 -> 121 [ label = "" ]
    63 -> 120 [ label = "" ]
    63 -> 119 [ label = "" ]
    63 -> 118 [ label = "" ]
    63 -> 117 [ label = "" ]
    124 -> 14 [ label = "" ]
    124 -> 75 [ label = "" ]
    65 -> 14 [ label = "" ]
    65 -> 125 [ label = "" ]
    65 -> 126 [ label = "" ]
    65 -> 127 [ label = "" ]
    65 -> 124 [ label = "" ]
    65 -> 128 [ label = "" ]
    128 -> 14 [ label = "" ]
    128 -> 74 [ label = "" ]
    127 -> 14 [ label = "" ]
    127 -> 73 [ label = "" ]
    126 -> 14 [ label = "" ]
    126 -> 72 [ label = "" ]
    125 -> 14 [ label = "" ]
    125 -> 71 [ label = "" ]
    75 -> 15 [ label = "" ]
    74 -> 15 [ label = "" ]
    73 -> 15 [ label = "" ]
    72 -> 15 [ label = "" ]
    71 -> 15 [ label = "" ]
    66 -> 16 [ label = "" ]
    129 -> 22 [ label = "" ]
    80 -> 22 [ label = "" ]
    81 -> 22 [ label = "" ]
    81 -> 129 [ label = "" ]
    58 -> 25 [ label = "" ]
    86 -> 25 [ label = "" ]
    86 -> 58 [ label = "" ]
    69 -> 27 [ label = "" ]
    69 -> 130 [ label = "" ]
    131 -> 27 [ label = "" ]
    131 -> 130 [ label = "" ]
    131 -> 58 [ label = "" ]
    70 -> 27 [ label = "" ]
    70 -> 131 [ label = "" ]
    130 -> 27 [ label = "" ]
    84 -> 29 [ label = "" ]
    100 -> 29 [ label = "" ]
    100 -> 84 [ label = "" ]
    97 -> 30 [ label = "" ]
    97 -> 84 [ label = "" ]
    101 -> 30 [ label = "" ]
    101 -> 97 [ label = "" ]
    132 -> 31 [ label = "" ]
    132 -> 93 [ label = "" ]
    133 -> 31 [ label = "" ]
    133 -> 92 [ label = "" ]
    134 -> 31 [ label = "" ]
    134 -> 91 [ label = "" ]
    135 -> 31 [ label = "" ]
    135 -> 90 [ label = "" ]
    136 -> 31 [ label = "" ]
    136 -> 89 [ label = "" ]
    137 -> 31 [ label = "" ]
    137 -> 88 [ label = "" ]
    138 -> 31 [ label = "" ]
    138 -> 87 [ label = "" ]
    139 -> 31 [ label = "" ]
    139 -> 138 [ label = "" ]
    139 -> 137 [ label = "" ]
    139 -> 136 [ label = "" ]
    139 -> 135 [ label = "" ]
    139 -> 134 [ label = "" ]
    139 -> 133 [ label = "" ]
    139 -> 132 [ label = "" ]
    140 -> 31 [ label = "" ]
    140 -> 39 [ label = "" ]
    141 -> 31 [ label = "" ]
    142 -> 31 [ label = "" ]
    142 -> 143 [ label = "" ]
    142 -> 144 [ label = "" ]
    145 -> 31 [ label = "" ]
    146 -> 31 [ label = "" ]
    147 -> 31 [ label = "" ]
    147 -> 140 [ label = "" ]
    148 -> 31 [ label = "" ]
    148 -> 147 [ label = "" ]
    148 -> 146 [ label = "" ]
    148 -> 145 [ label = "" ]
    148 -> 142 [ label = "" ]
    144 -> 31 [ label = "" ]
    144 -> 25 [ label = "" ]
    64 -> 31 [ label = "" ]
    64 -> 141 [ label = "" ]
    64 -> 148 [ label = "" ]
    64 -> 139 [ label = "" ]
    143 -> 31 [ label = "" ]
    143 -> 0 [ label = "" ]
    93 -> 32 [ label = "" ]
    92 -> 32 [ label = "" ]
    91 -> 32 [ label = "" ]
    90 -> 32 [ label = "" ]
    89 -> 32 [ label = "" ]
    88 -> 32 [ label = "" ]
    87 -> 32 [ label = "" ]
    149 -> 33 [ label = "" ]
    149 -> 20 [ label = "" ]
    150 -> 33 [ label = "" ]
    150 -> 149 [ label = "" ]
    60 -> 33 [ label = "" ]
    60 -> 151 [ label = "" ]
    60 -> 150 [ label = "" ]
    151 -> 33 [ label = "" ]
    107 -> 38 [ label = "" ]
    152 -> 38 [ label = "" ]
    152 -> 30 [ label = "" ]
    153 -> 38 [ label = "" ]
    153 -> 84 [ label = "" ]
    153 -> 97 [ label = "" ]
    154 -> 38 [ label = "" ]
    154 -> 152 [ label = "" ]
    155 -> 38 [ label = "" ]
    102 -> 38 [ label = "" ]
    156 -> 38 [ label = "" ]
    103 -> 38 [ label = "" ]
    103 -> 156 [ label = "" ]
    103 -> 155 [ label = "" ]
    103 -> 154 [ label = "" ]
    103 -> 157 [ label = "" ]
    103 -> 153 [ label = "" ]
    157 -> 38 [ label = "" ]
    83 -> 41 [ label = "" ]
    158 -> 44 [ label = "" ]
    82 -> 44 [ label = "" ]
    82 -> 158 [ label = "" ]
    159 -> 45 [ label = "" ]
    160 -> 45 [ label = "" ]
    160 -> 98 [ label = "" ]
    76 -> 45 [ label = "" ]
    76 -> 160 [ label = "" ]
    76 -> 159 [ label = "" ]
    108 -> 46 [ label = "" ]
    98 -> 48 [ label = "" ]
    98 -> 106 [ label = "" ]
    106 -> 49 [ label = "" ]
    106 -> 108 [ label = "" ]
    105 -> 52 [ label = "" ]
    68 -> 52 [ label = "" ]
    104 -> 52 [ label = "" ]
    67 -> 52 [ label = "" ]
    95 -> 55 [ label = "" ]
    77 -> 55 [ label = "" ]
    94 -> 55 [ label = "" ]
    79 -> 55 [ label = "" ]
    78 -> 55 [ label = "" ]
}

